zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1088 @@
+            // and not subject to being reorg'ed.
+            let stable_height = new_tip.saturating_sub(PRUNING_DEPTH);
+
+            // if the wallet's prior tip is above the stable height, prioritize the range between
+            // it and the new tip as `ChainTip`. Otherwise, prioritize the `VALIDATION_DEPTH`
+            // blocks above the wallet's prior tip as `Verify`. Since `scan_cached_blocks`
+            // retrieves the metadata for the block being connected to, the connectivity to the
+            // prior tip will always be checked. Since `Verify` ranges have maximum priority, even
+            // if the block source begins downloading blocks from the shard scan range (which ends
+            // at the stable height) the scanner should not attempt to scan those blocks until the
+            // tip range has been completely checked and any required rewinds have been performed.
+            if prior_tip >= stable_height {
+                // This may overlap the `shard_entry` range and if so will be coalesced with it.
+                ScanRange::from_parts(prior_tip..chain_end, ScanPriority::ChainTip)
+            } else {
+                // The prior tip is in the range that we now expect to be stable, so we need to verify
+                // and advance it up to at most the stable height. The shard entry will then cover
+                // the range to the new tip at the lower `ChainTip` priority.
+                ScanRange::from_parts(
+                    prior_tip..min(stable_height, prior_tip + VALIDATION_DEPTH),
+                    ScanPriority::Verify,
+                )
+            }
+        }
+    });
+    if let Some(entry) = &shard_entry {
+        debug!("{} will update latest shard", entry);
+    }
+    if let Some(entry) = &tip_entry {
+        debug!("{} will connect prior tip to new tip", entry);
+    }
+
+    let query_range = match (shard_entry.as_ref(), tip_entry.as_ref()) {
+        (Some(se), Some(te)) => Some(Range {
+            start: min(se.block_range().start, te.block_range().start),
+            end: max(se.block_range().end, te.block_range().end),
+        }),
+        (Some(se), None) => Some(se.block_range().clone()),
+        (None, Some(te)) => Some(te.block_range().clone()),
+        (None, None) => None,
+    };
+
+    if let Some(query_range) = query_range {
+        replace_queue_entries(
+            conn,
+            &query_range,
+            shard_entry.into_iter().chain(tip_entry.into_iter()),
+        )?;
+    } else {
+        // If we have neither shard data nor any existing block data in the database, we should also
+        // have no existing scan queue entries and can fall back to linear scanning from Sapling
+        // activation.
+        if let Some(sapling_activation) = params.activation_height(NetworkUpgrade::Sapling) {
+            let scan_range =
+                ScanRange::from_parts(sapling_activation..chain_end, ScanPriority::Historic);
```

Comment:
How do we know that `chain_end >= sapling_activation`?

Reply:
We "know" this because we know that both mainnet and testnet chain tips are after Sapling activation, so it is functionally invalid to call `update_chain_tip` with a height prior to Sapling activation. Whether we panic here or explicitly earlier doesn't really matter. But again, if we're in this scenario then all we can do is "nothing".

Reply:
Hmm, but incorrect information from the lightwalletd server shouldn't cause a panic, should it?

Reply:
Pushed a commit that does nothing in this scenario.